designs/disruption-controls.md

```diff
@@ -0,0 +1,118 @@
+
+These controls will include a new `Disruption` block in the `v1beta1` `NodePool` that will contain all scale-down behavioral fields. Initially, this will include `Budgets`, which define a (1) parallelism of how many nodes can be deprovisioned at a time and a (2) cron schedule that determines when the parallelism applies, and a `ConsolidateAfter` field that will allow users to affect the speed at which Karpenter will scale down underutilized nodes. Future scale-down behavioral fields should be colocated with these fields.
+
+## Proposed Spec
+
+```yaml
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: default
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    consolidationPolicy: WhenUnderutilized || WhenEmpty
+    consolidateAfter: 10m || Never # metav1.Duration
+    expireAfter: 10m || Never # Equivalent to v1alpha5 TTLSecondsUntilExpired
+    budgets:
+    # On Weekdays during business hours, don't do any deprovisioning.
+    - crontab: "0 9 * * mon-fri"
+      duration: 8h
+      maxUnavailable: 0
+    # Every other time, only allow 10 nodes to be deprovisioned simultaneously
+    - maxUnavailable: 10
+```
+
+## Code Definition
+
+```go
+type Disruption struct {
+    {...}
+    // Budgets is a list of Budgets.
+    // If there are multiple active budgets, the most restrictive budget's maxUnavailable is respected.
+    Budgets []Budget `json:"budgets,omitempty" hash:"ignore"`
+    // ConsolidateAfter is a nillable duration, parsed as a metav1.Duration.
+    // Users can use "Never" to disable Consolidation.
+    ConsolidateAfter *NillableDuration `json:"consolidateAfter" hash:"ignore"`
+    // ConsolidationPolicy determines how Karpenter will consider nodes
+    // as candidates for Consolidation.
+    // WhenEmpty uses the same behavior as v1alpha5 TTLSecondsAfterEmpty
+    // WhenUnderutilized uses the same behavior as v1alpha5.ConsolidationEnabled: true
+    ConsolidationPolicy string `json:"consolidationPolicy" hash:"ignore"`
+}
+// Budget specifies periods of times where Karpenter will restrict the
+// number of Node Claims that can be terminated at a time.
+// Unless specified, a budget is always active.
+type Budget struct {
+    // MaxUnavailable dictates how many Node Claims owned by this NodePool
+    // can be terminating at once.
+    // This only respects and considers nodes with a DeletionTimestamp set.
+    MaxUnavailable intstr.IntOrString `json:"maxUnavailable" hash:"ignore"`
```

Comment:
Currently drift will only rotate one node at a time if it has a workload. 

Will this end up affecting drift and the number of nodes containing a workload which it can rotate in parallel?

Reply:
Yep! I'm working on pulling out some of the synchronous waits we do in code to enable Karpenter to drift/expire/consolidate multiple nodes at a time.